Interlace log lines with task progress

eatkins opened the pull request:
With the current supershell implementation, the progress display
flickers when there is heavy console logging during task evaluation.
This is because the console appender clears out the task progress and it
isn't restored until the next periodic super shell report (which
runs every 100ms by default). To remove the flickering, I reworked the
implementation to interlace the log lines with progress reports. In
order to ensure that the log lines remained contiguous, I had to apply
padding at the bottom of the supershell region whenever the new report
contained fewer lines than the old report. The report shifts down as new
log lines are appended. This isn't optimal, but I think removing
the flickering while preserving contiguous log lines is worth it.

eatkins committed:
Interlace log lines with task progress

With the current supershell implementation, the progress display
flickers when there is heavy console logging during task evaluation.
This is because the console appender clears out the task progress and it
isn't restored until the next periodic super shell report (which
runs every 100ms by default). To remove the flickering, I reworked the
implementation to interlace the log lines with progress reports. In
order to ensure that the log lines remained contiguous, I had to apply
padding at the bottom of the supershell region whenever the new report
contained fewer lines than the old report. The report shifts down as new
log lines are appended. This isn't optimal, but I think removing
the flickering while preserving contiguous log lines is worth it.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=develop&io=develop&librarymanagement=develop&util=pull/216/head&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eatkins commented:
I generated a before and after in the io project. Here is the result with RC5:
[![asciicast](https://asciinema.org/a/OSEfa0UAR5uWFSojhnwecbiCz.svg)](https://asciinema.org/a/OSEfa0UAR5uWFSojhnwecbiCz)
Notice how while the test lines are printing, the progress reports will disappear at times. Compare to the Snapshot version where the progress reports are more stable:
[![asciicast](https://asciinema.org/a/1JFbpTsjNzK6iIwXCHjZaqz1Y.svg)](https://asciinema.org/a/1JFbpTsjNzK6iIwXCHjZaqz1Y)

eatkins committed:
Manage progress padding

With this commit, I improved the padding management so that padding is
now added above the progress report. Whenever a line is logged at the
info or greater level, we can reduce the padding level by one since that
line has effectively filled in the padding.

eatkins commented:
I figured out how to anchor things on the bottom with a minor hack in https://github.com/sbt/util/pull/216/commits/a5666e97b6e2395c7404d21c85f3efc58778df19 :
[![asciicast](https://asciinema.org/a/RJYfLERhRmVcFidbtsflgwa2Z.svg)](https://asciinema.org/a/RJYfLERhRmVcFidbtsflgwa2Z)

eatkins committed:
Allow supershell to work in no color mode

Supershell actually works quite well in no color mode. On the sbt side,
we still want to disable supershell automatically if the output is not a
terminal or no color is set, but this commit allows the user to force
supershell through -Dsbt.supershell or the useSuperShell setting even
when no color is set.

eed3si9n reviewed (approved):
Thanks for looking into this!

eed3si9n commented on internal/util-logging/src/main/scala/sbt/internal/util/ConsoleAppender.scala:
```diff
@@ -311,6 +310,74 @@ object ConsoleAppender {
+      if (progress.length > 0) {
+        val stripped = ascii.matcher(l).replaceAll("")
+        val isDebugLine =
+          l.startsWith("[debug]") || l.startsWith(s"${scala.Console.RESET}[debug]")
```
Doesn't this depend on the log level of the logger?
